Харин Матвей. Вариант 9. Технология OMP. Вычисление многомерных интегралов с использованием многошаговой схемы (метод прямоугольников).

mateusxap opened the pull request:
Реализация представляет последовательное вычисление многомерных интегралов методом прямоугольников. Код принимает значения функции на равномерной сетке произвольной размерности, размеры сетки по каждому измерению и шаги интегрирования, а затем вычисляет приближённый интеграл как сумму всех значений функции, умноженную на элемент объёма (произведение шагов). Использование типа double обеспечивает высокую точность вычислений, а структура поддерживает любые размерности благодаря динамическому определению сетки и шагов.

Процесс разбит на этапы: PreProcessingImpl извлекает входные данные и проверяет их согласованность, ValidationImpl контролирует корректность структуры входов и выходов, RunImpl выполняет основное вычисление суммы и объёма, а PostProcessingImpl записывает результат.

В программе распараллеливание с помощью OpenMP реализовано через директиву #pragma omp parallel for с модификаторами reduction, которая распределяет итерации циклов между потоками для параллельного выполнения: в PreProcessingImpl вычисляется произведение размеров сетки (total_size) и проверка шагов (is_valid) с reduction(*:total_size) и reduction(&&:is_valid) соответственно, а в RunImpl параллельно суммируются значения функции (total) и вычисляется произведение шагов (volume_element) с reduction(+:total) и reduction(*:volume_element)

ValeraDanger commented on tasks/omp/kharin_m_multidimensional_integral_calc/src/ops_seq.cpp:
```diff
@@ -0,0 +1,105 @@
+  auto* steps_ptr = reinterpret_cast<double*>(task_data->inputs[2]);
+  step_sizes_ = std::vector<double>(steps_ptr, steps_ptr + d);
+
+  //   // Проверка на отрицательные шаги
```
remove comments

mateusxap replied:
fixed

ValeraDanger commented on tasks/omp/kharin_m_multidimensional_integral_calc/src/ops_seq.cpp:
```diff
@@ -0,0 +1,105 @@
+    total += input_[i];
+  }
+
+  // for (const auto& val : input_) {
```
remove comments

mateusxap replied:
fixed

ValeraDanger commented on tasks/omp/kharin_m_multidimensional_integral_calc/src/ops_seq.cpp:
```diff
@@ -0,0 +1,105 @@
+    volume_element *= step_sizes_[i];
+  }
+
+  // for (const auto& h : step_sizes_) {
```
remove comments

sk1er52 commented on tasks/omp/kharin_m_multidimensional_integral_calc/include/ops_seq.hpp:
```diff
@@ -0,0 +1,28 @@
+#pragma once
```
rename this file, it's not seq version

mateusxap replied:
fixed

sk1er52 commented on tasks/omp/kharin_m_multidimensional_integral_calc/src/ops_seq.cpp:
```diff
@@ -0,0 +1,87 @@
+#include "seq/kharin_m_multidimensional_integral_calc/include/ops_seq.hpp"
```
rename this file, it's not seq version

mateusxap replied:
fixed

sk1er52 commented on tasks/seq/kharin_m_multidimensional_integral_calc/func_tests/main.cpp:
```diff
@@ -0,0 +1,277 @@
+#include <gtest/gtest.h>
```
it's PR for OMP version, please remove seq version files

mateusxap replied:
fixed

aobolensk reviewed (changes requested):
Please, modify existing "_disabled" implementation. Also, note what has been changed in the fix.

mateusxap commented:
> Please, modify existing "_disabled" implementation. Also, note what has been changed in the fix.

Fixed: Renamed the namespace to _omp and the class to TestTaskOpenMP in the OMP implementation, and updated the tests accordingly.
